File: articles/life-death-problem-of-biological-definition/python/borderline_case_scoring.py
"""
Borderline-case scoring workflow.

Run:
    python python/borderline_case_scoring.py
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd

from definition_core import heuristic_life_score


ARTICLE_DIR = Path(__file__).resolve().parents[1]
CASES_PATH = ARTICLE_DIR / "data" / "borderline_cases.csv"
WEIGHTS_PATH = ARTICLE_DIR / "data" / "life_criteria_weights.csv"


def category(score: float) -> str:
    """Classify heuristic score without claiming a universal definition."""

    if score >= 0.72:
        return "strongly_life_like_under_this_matrix"
    if score >= 0.45:
        return "borderline_or_context_dependent"
    return "weakly_life_like_under_this_matrix"


def main() -> None:
    """Score borderline cases using explicit weights."""

    cases = pd.read_csv(CASES_PATH)
    weights_df = pd.read_csv(WEIGHTS_PATH)
    weights = dict(zip(weights_df["criterion"], weights_df["weight"]))

    score_rows = []

    for _, row in cases.iterrows():
        criteria = {criterion: float(row[criterion]) for criterion in weights}
        score = heuristic_life_score(criteria, weights)

        score_rows.append(
            {
                "case": row["case"],
                "heuristic_life_score": score,
                "category": category(score),
                **criteria,
            }
        )

    result = pd.DataFrame(score_rows).sort_values("heuristic_life_score", ascending=False)

    print(result.round(3).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/life-death-problem-of-biological-definition/r/borderline_case_scoring.R
# Borderline life-criteria matrix in R.

cases_path <- file.path("data", "borderline_cases.csv")
weights_path <- file.path("data", "life_criteria_weights.csv")

if (!file.exists(cases_path)) {
  cases_path <- file.path("..", "data", "borderline_cases.csv")
  weights_path <- file.path("..", "data", "life_criteria_weights.csv")
}

cases <- read.csv(cases_path)
weights_df <- read.csv(weights_path)

weights <- setNames(weights_df$weight, weights_df$criterion)

cases$heuristic_life_score <-
  cases$organization * weights[["organization"]] +
  cases$metabolism * weights[["metabolism"]] +
  cases$autonomy * weights[["autonomy"]] +
  cases$heredity * weights[["heredity"]] +
  cases$responsiveness * weights[["responsiveness"]] +
  cases$evolutionary_capacity * weights[["evolutionary_capacity"]]

cases$category <- ifelse(
  cases$heuristic_life_score >= 0.72,
  "strongly_life_like_under_this_matrix",
  ifelse(
    cases$heuristic_life_score >= 0.45,
    "borderline_or_context_dependent",
    "weakly_life_like_under_this_matrix"
  )
)

cases <- cases[order(-cases$heuristic_life_score), ]

print(round(cases, 3))
